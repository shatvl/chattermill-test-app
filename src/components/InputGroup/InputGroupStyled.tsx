import styled from 'styled-components';

export const InputGroupControl = styled.div`
  position: relative;
  margin-bottom: 1.5rem;
  width: 100%;
`;

export const InputGroupInput = styled.input<{ focused?: boolean }>`
  display: block;
  width: 100%;
  height: 3.375rem;
  padding: ${({ theme }) => `calc(.75rem - ${theme.borderWidth}) calc(.9rem - ${theme.borderWidth})`};
  border: ${({ theme }) => theme.borderWidth} solid ${({ theme }) => theme.colors.gray};
  border-radius: 0.25rem;
  color: ${({ theme }) => theme.text};
  outline: none;
  transition: border-color 0.15s ease;
  word-break: break-word;
  font-size: 1rem;
  line-height: 1.25rem;
  ::placeholder {
    opacity: ${({ focused }) => (focused ? '0' : '1')};
  }
`;

export const InputGroupLabel = styled.label<{ focused?: boolean; empty?: boolean }>`
  display: block;
  padding: 0 0.25rem;
  position: absolute;
  left: 0.7rem;
  top: -0.5rem;
  background-color: ${({ theme }) => theme.colors.background};
  color: ${({ theme }) => theme.colors.textSecondary};
  font-size: 0.75rem;
  font-weight: 400;
  opacity: ${({ focused, empty }) => (focused || !empty ? 1 : 0)};
  transition: all 0.15s ease;
`;
