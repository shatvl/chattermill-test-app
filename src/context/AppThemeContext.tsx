import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
