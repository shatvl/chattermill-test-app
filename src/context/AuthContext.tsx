import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
