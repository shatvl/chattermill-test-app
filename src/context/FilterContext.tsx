import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
