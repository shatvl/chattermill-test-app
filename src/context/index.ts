export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
