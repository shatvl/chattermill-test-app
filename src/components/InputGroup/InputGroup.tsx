import React, { FC, useState } from 'react';
import { InputGroupLabel, InputGroupControl, InputGroupInput } from './InputGroupStyled';

export enum InputTypes {
  TEXT = 'text',
  PASSWORD = 'password',
  NUMBER = 'number',
}

interface InputGroupProps {
  label?: string;
  value: string;
  formControlName: string;
  onChange: (key: string, value: string) => void;
  type?: InputTypes;
  placeholder?: string;
}

const InputGroup: FC<InputGroupProps> = ({ formControlName, type = InputTypes.TEXT, value, onChange, label, placeholder }) => {
  const [focused, setFocused] = useState(false);

  const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => onChange(formControlName, event.target.value);

  return (
    <InputGroupControl>
      <InputGroupInput
        type={type}
        value={value}
        focused={focused}
        onChange={handleOnChange}
        onFocus={() => setFocused(true)}
        onBlur={() => setFocused(false)}
        placeholder={placeholder ?? label}
      />
      {label && (
        <InputGroupLabel empty={!value} focused={focused}>
          {label}
        </InputGroupLabel>
      )}
    </InputGroupControl>
  );
};

export default InputGroup;
