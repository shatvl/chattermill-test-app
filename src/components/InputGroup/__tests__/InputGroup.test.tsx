/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { mount } from 'enzyme';
import { wait, waitRenderer } from '../../../testing/wait';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import InputGroup from '../InputGroup';
import { InputGroupInput, InputGroupLabel } from '../InputGroupStyled';

describe('renders input group', () => {
  it('input group matches snapshot', async () => {
    const onChangeFn = jest.fn();

    const tree = renderer.create(
      <AppThemeProvider>
        <InputGroup onChange={onChangeFn} value="" formControlName="testfield" />
      </AppThemeProvider>,
    );

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('check onchange in input group', async () => {
    const onChangeFn = jest.fn();
    const mockedEvent = { target: { name: 'change', value: 'testing?' } };
    const mockedControlName = 'testfield';

    const component = mount(
      <AppThemeProvider>
        <InputGroup onChange={onChangeFn} value="" formControlName={mockedControlName} />
      </AppThemeProvider>,
    );

    await wait();

    component.update();

    component.find(InputGroupInput).simulate('change', mockedEvent);

    expect(onChangeFn).toBeCalledWith(mockedControlName, mockedEvent.target.value);
  });

  it('check "floating" label', async () => {
    const onChangeFn = jest.fn();
    const label = 'label';

    const component = mount(
      <AppThemeProvider>
        <InputGroup label={label} onChange={onChangeFn} value="" formControlName="testfield" />
      </AppThemeProvider>,
    );

    await wait();

    component.update();

    expect(component.find(InputGroupLabel).prop('focused')).toBeFalsy();
    expect(component.find(InputGroupLabel).prop('children')).toEqual(label);

    //focus input and label will be visible (focused)
    component.find(InputGroupInput).simulate('focus');
    component.update();
    expect(component.find(InputGroupLabel).prop('focused')).toBeTruthy();

    //blur input and label will be invisible
    component.find(InputGroupInput).simulate('blur');
    expect(component.find(InputGroupLabel).prop('focused')).toBeFalsy();
  });
});
