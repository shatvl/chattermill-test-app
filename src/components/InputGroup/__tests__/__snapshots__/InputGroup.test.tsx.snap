// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders input group input group matches snapshot 1`] = `
.c0 {
  position: relative;
  margin-bottom: 1.5rem;
  width: 100%;
}

.c1 {
  display: block;
  width: 100%;
  height: 3.375rem;
  padding: calc(.75rem - 0.0625rem) calc(.9rem - 0.0625rem);
  border: 0.0625rem solid #c4c9cc;
  border-radius: 0.25rem;
  outline: none;
  -webkit-transition: border-color 0.15s ease;
  transition: border-color 0.15s ease;
  word-break: break-word;
  font-size: 1rem;
  line-height: 1.25rem;
}

.c1::-webkit-input-placeholder {
  opacity: 1;
}

.c1::-moz-placeholder {
  opacity: 1;
}

.c1:-ms-input-placeholder {
  opacity: 1;
}

.c1::placeholder {
  opacity: 1;
}

<div
  className="c0"
>
  <input
    className="c1"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
</div>
`;
